cpp/src/arrow/util/converter.h

```diff
@@ -0,0 +1,281 @@
+    return this->builder_->Reserve(additional_capacity);
+  }
+
+  Status AppendNull() override { return this->builder_->AppendNull(); }
+
+  Result<std::shared_ptr<Array>> Finish() override { return builder_->Finish(); };
+
+ protected:
+  const T& type_;
+  BuilderType* builder_;
+};
+
+template <typename T, typename BaseConverter>
+class PrimitiveArrayConverter : public TypedArrayConverter<T, BaseConverter> {
+ public:
+  using TypedArrayConverter<T, BaseConverter>::TypedArrayConverter;
+};
+
+template <typename T, typename BaseConverter>
+class DictionaryArrayConverter
+    : public TypedArrayConverter<DictionaryType, BaseConverter, DictionaryBuilder<T>> {
+ public:
+  DictionaryArrayConverter(const std::shared_ptr<DataType>& type,
+                           std::shared_ptr<ArrayBuilder> builder,
+                           typename BaseConverter::OptionsType options)
+      : TypedArrayConverter<DictionaryType, BaseConverter, DictionaryBuilder<T>>(
+            type, builder, options),
+        value_type_(checked_cast<const T&>(
+            *checked_cast<const DictionaryType&>(*type).value_type())) {}
+
+ protected:
+  const T& value_type_;
+};
+
+template <typename T, typename BaseConverter>
+class ListArrayConverter : public TypedArrayConverter<T, BaseConverter> {
+ public:
+  ListArrayConverter(const std::shared_ptr<DataType>& type,
+                     std::shared_ptr<ArrayBuilder> builder,
+                     std::shared_ptr<BaseConverter> value_converter,
+                     typename BaseConverter::OptionsType options)
+      : TypedArrayConverter<T, BaseConverter>(type, builder, options),
+        value_converter_(std::move(value_converter)) {}
+
+ protected:
+  std::shared_ptr<BaseConverter> value_converter_;
+};
+
+template <typename T, typename BaseConverter>
+class StructArrayConverter : public TypedArrayConverter<T, BaseConverter> {
+ public:
+  StructArrayConverter(const std::shared_ptr<DataType>& type,
+                       std::shared_ptr<ArrayBuilder> builder,
+                       std::vector<std::shared_ptr<BaseConverter>> child_converters,
+                       typename BaseConverter::OptionsType options)
+      : TypedArrayConverter<T, BaseConverter>(type, builder, options),
+        child_converters_(std::move(child_converters)) {}
+
+ protected:
+  std::vector<std::shared_ptr<BaseConverter>> child_converters_;
+};
+
+#define DICTIONARY_CASE(TYPE_ENUM, TYPE_CLASS)                                          \
+  case Type::TYPE_ENUM:                                                                 \
+    out->reset(new DictionaryConverter<TYPE_CLASS>(type, std::move(builder), options)); \
+    break;
+
+template <typename Options, typename BaseConverter,
+          template <typename...> class PrimitiveConverter,
+          template <typename...> class DictionaryConverter,
+          template <typename...> class ListConverter,
+          template <typename...> class StructConverter>
```

Comment:
I'd like to keep the construction code somewhere else than `python_to_arrow.cc`. I moved it into the internal namespace so we can change it anytime. 

I could reduce the four converter kinds to two by specializing TypedConverter for list-like and struct type, but not sure how much would it help. 

The previous code was using static dispatch extensively so I tried to keep it, we could certainly simplify the API by using type visitors but I'm afraid that would cause performance regressions.